midenc-tx-kernel/Cargo.toml

```diff
@@ -0,0 +1,18 @@
+[package]
+name = "midenc-tx-kernel"
```

Comment:
My understanding was that we were going to name wrapper crates like this something to the effect of `miden-<name>-sys`.
where `<name>` is something like `stdlib`, `tx_kernel`, etc.

While this crate doesn't contain bindings yet, I would expect that whatever crate _does_ contain bindings (or is expected to eventually), would also be responsible for producing the `Library` (and embedding it in the crate for distribution).

The Miden standard library is a bit weird at the moment, as it is currently maintained in `miden-vm`, and the `miden-stdlib` crate does not contain bindings, instead we have `-sys` crates in this repo which handle that, but the `Library` is embedded in `miden-stdlib`. But I wouldn't pay much mind to that - instead, we should define `miden-base-sys`, and have that define a `Library` containing the code from `miden-base` (or our stubs, as the case may be for the moment). We would then link it using `-l base`.

I'm not sure we need to have separate crates for each subcomponent of `miden-base`, since it is basically all-or-nothing, but I'm open to reconsidering that aspect.

Reply:
> My understanding was that we were going to name wrapper crates like this something to the effect of `miden-<name>-sys`. where `<name>` is something like `stdlib`, `tx_kernel`, etc.

Yes, but I have not connected the dots and started a new crate. 🤦 

> While this crate doesn't contain bindings yet, I would expect that whatever crate _does_ contain bindings (or is expected to eventually), would also be responsible for producing the `Library` (and embedding it in the crate for distribution).

Makes sense. Does it mean that we no longer need to separately specify digests for each function in the bindings? Like in https://github.com/0xPolygonMiden/compiler/blob/ab2b7257965b0f6e56d75e84f403d3b9fd683c4b/sdk/tx-kernel-sys/src/externs.rs#L6-L8

> The Miden standard library is a bit weird at the moment, as it is currently maintained in `miden-vm`, and the `miden-stdlib` crate does not contain bindings, instead we have `-sys` crates in this repo which handle that, but the `Library` is embedded in `miden-stdlib`. But I wouldn't pay much mind to that - instead, we should define `miden-base-sys`, and have that define a `Library` containing the code from `miden-base` (or our stubs, as the case may be for the moment). We would then link it using `-l base`.
> 
> I'm not sure we need to have separate crates for each subcomponent of `miden-base`, since it is basically all-or-nothing, but I'm open to reconsidering that aspect.

I agree.

Reply:
> Makes sense. Does it mean that we no longer need to separately specify digests for each function in the bindings?

My thinking was that the build script for the crate would compile the MAST, write that to `OUT_DIR` (and then store it in the crate sources using `include_bytes!`), as well as generate a Rust source file containing the extern function definitions for the procedures in the MAST, using the digests that we can get from the `Library` struct by visiting each module and procedure, paired with signature metadata that we get initially from something we maintain by hand (maybe JSON? doesn't matter, whatever is most convenient, could also add a dependency on `midenc-hir-type` and serialize types to/from JSON), but when I am able to land the changes that add type signatures to Miden Assembly, we would derive the signature metadata from that. So essentially we are able to automate the generation of bindings for the most part for `-sys` crates.

The aspect that would be hand-maintained, would be the high-level Rust crate that depends on the `-sys` crate - these would provide more native abstractions for the underlying raw procedure bindings.

This gets us the ability to load MAST from a Rust crate in the compiler (or in the test suite, etc.), like we do with `miden-stdlib`, the ability to generate the bindings is at nice benefit as well. We may want to use a separate process for managing the compilation of the MAST, and simply read the existing MAST to generate the bindings, but for now this will simplify some things.